torch/distributed/device_mesh.py

```diff
@@ -476,21 +483,28 @@ def _get_or_create_default_group(self):
                     f"Mesh should not be bigger than default world size {world_size}, but found {self.mesh.numel()} ranks!"
                 )
 
+            # ONLY set the device if the current device is not initialized, if user already
+            # set the device before DeviceMesh init, we respect the user's choice.
             device_handle = _get_device_handle(self.device_type)
-            # TODO: if user want to pass pg_options, offer a way to do it
-            if not default_initialized and device_handle:
-                # automatically set the current cuda/cuda-like device base on num of gpu devices available in each host
-                # NOTE: This device selection would only work for homogeneous hardware.
-                num_devices_per_host = device_handle.device_count()
-                if (
-                    world_size > num_devices_per_host
-                    and world_size % num_devices_per_host != 0
-                ):
-                    raise RuntimeError(
-                        f"DeviceMesh only support homogeneous hardware, but found "
-                        f"{world_size} ranks and {num_devices_per_host} {self.device_type} devices!"
-                    )
-                device_handle.set_device(get_rank() % num_devices_per_host)
+            if device_handle and not device_handle.is_initialized():
```

Comment:
I'm not familiar with the code here. Could you elaborate a bit what are being done here?
It looks like
- before this PR, the code would be executed if `device_handle` exists and c10d `is_initialized()` is false.
- after this PR, the code would be executed if `device_handle` exists and `device_handle`'s `is_initialized()` is false.

How is this related to
> when user did not set the device before calling the DeviceMesh constructor

Reply:
This `is_initialized()` is either `torch.cuda.is_initialized()` or `torch.cpu.is_initialized()`. For CUDA, this means the device state is initialized, iiuc.

Reply:
Yes, I also updated the PR summary to reflect the detailed changes this PR did.

Reply:
according to your PR desc, we don't do this `set_device`
> If user call init_process_group to init a world process group and the default cuda context is initialized

To translate it into this condition, wouldn't it be?
```suggestion
            if not default_initialized or (device_handle and not device_handle.is_initialized()):
```

Reply:
Oh good catch! The reason I did not check for `default_initialized` here anymore is that:
1. if user init a process group and bind a device (after binding a device it would initialized the cuda device state), this path would not be hit as `device_handle.is_initialized=True`
2. if user init a process group but forgot to bind a device (i.e. `not device_handle.is_initialized`), we can either 1) just silently not do anything and rely on user to set the device later after the DeviceMesh initialization, this is actually the behavior before this change 2) we still try to set up the device for users. I was thinking 2) is better in some sense, but yeah maybe we should still keep the behaviors of 1)

Reply:
updated with the changes